proposed/0007-Address_relocation.md

```diff
@@ -0,0 +1,40 @@
+- Feature Name: Prevent any single address being used more than once during address relocation.
+- Type enhancement
+- Related components routing
+- Start Date: 21-09-2015
+- RFC PR: (leave this empty)
+- Issue number: (leave this empty)
+
+# Summary
+
+Prevent a potential attack where a node gains overall/majority influence in a group setting.
+
+# Motivation
+
+The possibility 'may' exist for a node on the network to generate off-line derived identities close to a specific target and to then connect nodes whose resultant network relocated identities actually lie close to that target. In such cases the owner of that node will dominate, or at least have a dominant influence within, the group close to that target. This RFC proposes an amendment to the current process to prevent this possibility arising in practice.
+
+# Detailed design
+
+A node joining the SAFE network generates a new cryptographic key pair to access the network for each session. The generated public key, K, is then cryptographically hashed using some hash function H to give H(K). The node then sends a message, containing K, to the group H(K) requesting a network identifier for the session. On receipt of such a request each node belonging to the group hashes the concatenation of K with the two closest nodes, N1, N2 say, in xor distance from K present in their routing tables, giving H(K + N1 + N2) = N, the identifier the node will use on the network for the session.
+
+Having computed N, each node in the group H(K) sends a message, containing K, to the close group of N to confirm address relocation has been applied for the node owning K. A sufficient condition for the receiving group to resolve the message is for a quorum in the sending group to agree on the two closest nodes. Each node in the group H(K) further sends a message, containing N, to the joining node confirming the network address it now occupies.
+
+In it's current form the process ensures that a joining node cannot occupy a position of it's own choosing. However, in a semi-stable network state it may be possible given the joining node knows the two closest nodes, N3, N4 say, in it’s joined group to generate keys that are relocated close to some identifier, I say, and thus obtain group influence at that location. It is the purpose of this RFC to avoid that possibility by storing the two close node identities, N3 and N4, at each node belonging to the joined group in order to reject relocations involving either of those nodes if they occur more than once for the participating nodes in a given session.
```

Comment:
what does `semi-stable network state` mean?
Say an attacker having the first faulty node joined (via closing peer N1&N2 and K1) and  knows a closing peer N3&N4, then he can starts calculating a key so that `H(K + N3 + N4)` will be close to the target.
I understand the paragraph here is trying to avoid such attack scenario, but the chance that N3&N4 had been used previously (so get stored in the targeting group ) is no higher than the normal joining procedure. So I don't see the point how it can prevent such attack and how a normal joining procedure won't get blocked by this proposal.